api/base/filters.py

```diff
@@ -20,53 +28,128 @@ def filter_queryset(self, request, queryset, view):
             return queryset.sort(*ordering)
         return queryset
 
-query_pattern = re.compile(r'filter\[\s*(?P<field>\S*)\s*\]\s*')
-
-
-def query_params_to_fields(query_params):
-    return {
-        query_pattern.match(key).groupdict()['field']: value
-        for key, value in query_params.items()
-        if query_pattern.match(key)
-    }
-
 
 class FilterMixin(object):
     """ View mixin with helper functions for filtering. """
 
+    QUERY_PATTERN = re.compile(r'^filter\[(?P<field>\w+)\](\[(?P<op>\w+)\])?$')
+    COMPARISON_OPERATORS = ('gt', 'gte', 'lt', 'lte')
+    COMPARABLE_FIELDS = (ser.DateField, ser.DateTimeField, ser.DecimalField, ser.IntegerField)
+    MATCH_OPERATORS = ('contains', 'icontains')
+    MATCHABLE_FIELDS = (ser.CharField, ser.ListField)
     DEFAULT_OPERATOR = 'eq'
+    DEFAULT_OPERATOR_OVERRIDES = {
+        ser.CharField: 'icontains',
+        ser.ListField: 'contains',
+    }
+
+    TIME_FIELDS = (ser.DateTimeField, ser.DateField)
+    DATETIME_PATTERN = re.compile(r'^\d{4}\-\d{2}\-\d{2}(?P<time>T\d{2}:\d{2}(:\d{2}(:\d{1,6})?)?)$')
 
     def __init__(self, *args, **kwargs):
         super(FilterMixin, self).__init__(*args, **kwargs)
         if not self.serializer_class:
             raise NotImplementedError()
 
-    def is_filterable_field(self, key):
-        try:
-            return key.strip() in self.serializer_class.filterable_fields
-        except AttributeError:
-            return key.strip() in self.serializer_class._declared_fields
-
-    # Used so that that queries by _id will work
-    def convert_key(self, key):
-        key = key.strip()
-        if self.serializer_class._declared_fields[key].source:
-            return self.serializer_class._declared_fields[key].source
-        return key
+    def _get_default_operator(self, field):
+        return self.DEFAULT_OPERATOR_OVERRIDES.get(type(field), self.DEFAULT_OPERATOR)
+
+    def parse_query_params(self, query_params):
+        """Maps query params to a dict useable for filtering
+        :param dict query_params:
+        :raises InvalidFilterError: If the filter field is not valid
+        :raises InvalidFilterComparisonType: If the query contains comparisons against non-date or non-numeric fields
+        :raises InvalidFilterMatchType: If the query contains comparisons against non-string or non-list fields
+        :raises InvalidFilterOperator: If the filter operator is not a member of self.COMPARISON_OPERATORS
+        :return dict: of the format {
+            <resolved_field_name>: {
+                'op': <comparison_operator>,
+                'value': <resolved_value>
+            }
+        }
+        """
+        fields = {}
+        for key, value in query_params.iteritems():
+            match = self.QUERY_PATTERN.match(key)
+            if match:
+                field_name = match.groupdict().get('field').strip()
+                if field_name not in self.serializer_class._declared_fields:
```

Comment:
On second look, `self.serializer_class` is a type, not an instance and `fields` is a property method